### ISSUE TITLE
vehicle_data endpoint charge_energy_added sometimes goes backwards

### ISSUE THREAD

**Issue by purcell-lab** [teslamotors/vehicle-command#224]
I sometimes see the charge_energy_added response going backwards when charging a vehicle.

Is this expected behaviour?

**Comment by jbanyer** [comment]
I have seen this happen when the vehicle is not charging: the charge_energy_added value can flap up and down slightly.

I haven't observed it when charging, but we don't really use the value so we wouldn't notice if it went backwards.

**Comment by purcell-lab** [comment]
Here's an example from the streaming entity during charging  going backwards.

ACChargingEnergyIn: 3.291, 3.270, 3.417

```log
data: {"data":{"ACChargingPower":"5.100","BMSState":"Charge","ACChargingEnergyIn":"3.144"},"createdAt":"2024-04-02T02:21:21.071986716Z","vin":"5YJ3F7EC5LF698201"}
data: {"data":{"ACChargingPower":"5.100","BMSState":"Charge","ACChargingEnergyIn":"3.270"},"createdAt":"2024-04-02T02:21:51.073026767Z","vin":"5YJ3F7EC5LF698201"}
data: {"data":{"ACChargingPower":"5.100","BMSState":"Charge","ACChargingEnergyIn":"3.291"},"createdAt":"2024-04-02T02:22:21.071057565Z","vin":"5YJ3F7EC5LF698201"}
data: {"data":{"ACChargingPower":"5.100","BMSState":"Charge","ACChargingEnergyIn":"3.270"},"createdAt":"2024-04-02T02:22:51.073411747Z","vin":"5YJ3F7EC5LF698201"}
data: {"data":{"ACChargingPower":"5.100","BMSState":"Charge","ACChargingEnergyIn":"3.417"},"createdAt":"2024-04-02T02:23:21.070641762Z","vin":"5YJ3F7EC5LF698201"}
data: {"data":{"ACChargingPower":"5.100","BMSState":"Charge","ACChargingEnergyIn":"3.417"},"createdAt":"2024-04-02T02:23:51.074006926Z","vin":"5YJ3F7EC5LF698201"}
data: {"data":{"ACChargingPower":"5.100","BMSState":"Charge","ACChargingEnergyIn":"3.417"},"createdAt":"2024-04-02T02:24:21.071414160Z","vin":"5YJ3F7EC5LF698201"}
data: {"data":{"ACChargingPower":"5.100","BMSState":"Charge","ACChargingEnergyIn":"3.417"},"createdAt":"2024-04-02T02:24:51.071301393Z","vin":"5YJ3F7EC5LF698201"}
data: {"data":{"ACChargingPower":"5.100","BMSState":"Charge","ACChargingEnergyIn":"3.542"},"createdAt":"2024-04-02T02:25:21.071561073Z","vin":"5YJ3F7EC5LF698201"}
```

**Comment by purcell-lab** [comment]
## Status update — May 2026

**Findings**

This is still open and unaddressed since 2024. There's been independent corroboration from multiple downstream integrations:

- [Teslemetry/hass-teslemetry#61](https://github.com/Teslemetry/hass-teslemetry/issues/61) — first reported April 2024, **still occurring as of July 2025** on the DC charging sensor (`sensor.*_dc_charging_energy_in`).
- [alandtse/tesla#886](https://github.com/alandtse/tesla/issues/886) — same pattern via teslajsonpy / Tesla Custom Integration.
- [home-assistant/core#112558](https://github.com/home-assistant/core/issues/112558) — surfaced as a core HA warning, escalating downstream pain.
- [Teslemetry/hass-teslemetry#37](https://github.com/Teslemetry/hass-teslemetry/issues/37) (closed) — Tessie and Teslemetry both seeing it.

**Characterisation of the bug** (consolidated from the threads):

- The value goes backwards by **20–200 Wh** during normal charging, then catches up on the next sample.
- It also fluctuates when the vehicle is **idle but warm** — battery thermal management, sentry mode, HVAC running on a completed session. Tomas's case in #61 is exactly this: HVAC on, charger session complete, no other sessions allowed by the charger → `dc_charging_energy_in` flapping.
- Affects both `ACChargingEnergyIn` (streaming telemetry) and `charge_energy_added` (Fleet API `vehicle_data` polling).

My streaming sample from April 2024:
```
ACChargingEnergyIn: 3.144 → 3.270 → 3.291 → 3.270 → 3.417 → 3.417 → 3.417 → 3.542
```

The 3.291 → 3.270 step (a 21 Wh decrease while `BMSState: Charge` and `ACChargingPower: 5.1 kW`) is the canonical bug — the vehicle is actively charging and the cumulative-energy counter still went down.

**Proposed way forward**

This is a Tesla-side data quality issue, but it's been costing downstream integrations real engineering effort for nearly two years. Asks that would unblock the ecosystem:

1. **Acknowledgement and root-cause** — even a comment confirming whether this is (a) sensor-noise that's expected to be filtered downstream, (b) a known firmware bug being tracked internally, or (c) a Fleet API serialisation issue would direct downstream work appropriately.
2. **If (a):** publish the tolerance band Tesla considers normal so downstream integrations can apply a consistent filter rather than each re-inventing their own threshold.
3. **If (b) or (c):** track to a firmware/API fix and let downstream integrations remove their workarounds once shipped.

Happy to provide more capture sessions across different vehicle states (charging / parked-warm / parked-cold / HVAC-on) if that helps narrow it down. CC @bre77, who's been carrying the downstream filter logic in hass-teslemetry.
